### ISSUE TITLE
Add `isEmpty()` check on built-in methods that take `String` as an argument

### ISSUE THREAD

**Issue by sjaakd** [mapstruct/mapstruct#441]
Like this:

```java
if ( date == null || date.trim().isEmpty()) {
            return null;
}
```

**Comment by sjaakd** [comment]
Hmm. looking at this.. 

I agree with the approach. 

But we should perhaps use a special dateformat (or dateformat=null) for indicating a construction by means of lexical representation (GYear, GMonth, dateTime.. etc. the stuff you find in XML) so:
`public abstract XMLGregorianCalendar newXMLGregorianCalendar(String lexicalRepresentation)`

dateformat=null, probably has the drawback that its not downwards compatible. So, should be realized before 1.0RC's

**Comment by gunnarmorling** [comment]
Probably it's just too late for me, but could you give an example for this:

> But we should perhaps use a special dateformat (or dateformat=null) for indicating a construction by means of lexical representation

?

**Comment by gunnarmorling** [comment]
Thinking about it, the same applies to all the implicit conversions from String.

I'm inclined to say we should handle empty strings equally to null in all these places. Plus, we may provide a mapper + global option to return to the current behavior: HANDLE_EMPTY_STRINGS_AS_NULL (defaulting to true).

WDYT?

**Comment by sjaakd** [comment]
> Probably it's just too late for me, but could you give an example for this:

I adapted the UT for this. Have a look :smile: 

But in general, the lexical representation of XML formats in [`XMLGregorianCalendar `](http://docs.oracle.com/javase/7/docs/api/javax/xml/datatype/DatatypeFactory.html#newXMLGregorianCalendar%28java.lang.String%29) differ from the ones allowed by [`SimpleDateFormat`](http://docs.oracle.com/javase/7/docs/api/java/text/SimpleDateFormat.html) (follow the hyper links). It feels strange that if we do not define a `@yyyMapping#dateFormat` we use the locale (by constructing a `SimpleDateFormat`) and then expect the date to comply to that locale. Although this is in line with other date conversions in our solution, it misses the primary use case. 

My considerations: That's why I mentioned we should perhaps 'use a special dateFormat' of `@yyyMapping#dateFormat` to allow a format 'XML', e.g. dateFormat="xml". Drawbacks:
1.  It does not make sense on other conversions / built-in concerning dates.
2. Absent `@yyyMapping#dateFormat` perhaps better reflects the natural `XmlGregorianCalendar`mapping, stemming from the W3 xmlschema definition.
3. Choosing a different format is not downwards compatible with earlier versions of MapStruct

**Comment by sjaakd** [comment]
> I'm inclined to say we should handle empty strings equally to null in all these places. Plus, we may provide a mapper + global option to return to the current behavior: HANDLE_EMPTY_STRINGS_AS_NULL (defaulting to true).

What places do you mean :smile: ? I thought there would be many (hence the original title) , but all of them are actually conversions. There's only one builtin that takes a `String` as input.



> Plus, we may provide a mapper + global option to return to the current behavior: HANDLE_EMPTY_STRINGS_AS_NULL (defaulting to true).
> 
> WDYT?

I think that's a good idea. But, lets address that in a separate issue. There's only one occurrence (yet). Feels like a bit too much. Lets see when the use case pops up.

**Comment by gunnarmorling** [comment]
> That's why I mentioned we should perhaps 'use a special dateFormat' of @yyyMapping#dateFormat to allow a format 'XML'

Ah, can't we automatically deduct the right type of format based on the fact whether a `@Mapping` is for an XML calendar or a standard `Date`, `Calendar` etc.? We could relax the contract/description of `@Mapping#dateFormat()` in that regard.

**Comment by gunnarmorling** [comment]
> What places do you mean :smile: ? I thought there would be many (hence the original title) , but all of them are actually conversions. 

Indeed it's built-in conversions I meant. E.g. from String to int. The generated code will raise a `NumberFormatException` atm. when passing in an empty string.

**Comment by sjaakd** [comment]
> Indeed it's built-in conversions I meant. E.g. from String to int. The generated code will raise a NumberFormatException atm. when passing in an empty string.

Ok. I'm splitting hairs now a bit. But to make sure that we're on par.  We've got:
1. Type Conversions
2. BuiltIn methods.

For 1. these are typically one liners generated in the type-conversion code.. For 2. we've got freemarker templates.

There's only one occurrence in 2. were we use a `String` as input argument. That's what I meant. 

Note: implementing an empty check for 1. is perhaps more tricky (e.g. ternary operator, driven by your proposed: `HANDLE_EMPTY_STRINGS_AS_NULL`

**Comment by sjaakd** [comment]
> Ah, can't we automatically deduct the right type of format based on the fact whether a @Mapping is for an XML calendar or a standard Date, Calendar etc.? We could relax the contract/description of @Mapping#dateFormat() in that regard.

You're there. That's exactly what I did (see PR #444 ). I need to update the documentation though..  (and add that this to the changed behaviour)..

**Comment by gunnarmorling** [comment]
> Ok. I'm splitting hairs now a bit. But to make sure that we're on par. We've got:
> 1. Type Conversions
> 2. BuiltIn methods.

Yes, I know :)

Bute note for a user that does not make a difference really, it's an implementation detail whether the conversion is done via an in-line assignment or a private method we generate. Thus behavior should be the same in bot hcases.

> Note: implementing an empty check for 1. is perhaps more tricky

Would it be more tricky really? The following should suffice IMO (controlled by the proposed option):

    if ( source.getFoo() != null && !source.getFoo().trim().isEmpty() ) {
        target.setFoo( Long.parseLong( source.getFoo() ) );
    }

**Comment by gunnarmorling** [comment]
> That's exactly what I did (see PR #444 ). I need to update the documentation though.. (and add that this to the changed behaviour)

+1

**Comment by gunnarmorling** [comment]
Btw. it seems this issue really is two independent ones. We should split them up.

**Comment by sjaakd** [comment]
> if ( source.getFoo() != null && !source.getFoo().trim().isEmpty() ) {
>     target.setFoo( Long.parseLong( source.getFoo() ) );
> }

Of course. Now I understand you. My confusion: the conversion string is now handled in the package: `org.mapstruct.ap.conversion` whereas the null check is in a wrapper.

> Btw. it seems this issue really is two independent ones. We should split them up.

You're right.. Just did: #449. I'll take the isEmpty checkout out of #449

**Comment by lukaszungier** [comment]
Is that problem still not solved? I ran into a related problem (#1505).
What's the latest status?
